File: swim-java/swim-core-java/swim.codec/src/main/java/swim/codec/OutputBuffer.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.codec;

import java.io.IOException;
import java.nio.channels.ReadableByteChannel;

/**
 * Non-blocking token stream buffer.
 */
public abstract class OutputBuffer<T> extends Output<T> {
  public abstract OutputBuffer<T> isPart(boolean isPart);

  public abstract int index();

  public abstract OutputBuffer<T> index(int index);

  public abstract int limit();

  public abstract OutputBuffer<T> limit(int limit);

  public abstract int capacity();

  public abstract int remaining();

  public abstract byte[] array();

  public abstract int arrayOffset();

  public abstract boolean has(int index);

  public abstract int get(int index);

  public abstract void set(int index, int token);

  public abstract int write(ReadableByteChannel channel) throws IOException;

  @Override
  public abstract OutputBuffer<T> write(int token);

  @Override
  public OutputBuffer<T> write(String string) {
    OutputBuffer<T> output = this;
    final int n = string.length();
    for (int i = 0; i < n; i = string.offsetByCodePoints(i, 1)) {
      output = output.write(string.codePointAt(i));
    }
    return output;
  }

  @Override
  public OutputBuffer<T> writeln(String string) {
    return write(string).writeln();
  }

  @Override
  public OutputBuffer<T> writeln() {
    return write(settings().lineSeparator());
  }

  @Override
  public OutputBuffer<T> display(Object object) {
    Format.display(object, this);
    return this;
  }

  @Override
  public OutputBuffer<T> debug(Object object) {
    Format.debug(object, this);
    return this;
  }

  public abstract OutputBuffer<T> move(int fromIndex, int toIndex, int length);

  public abstract OutputBuffer<T> step(int offset);

  @Override
  public OutputBuffer<T> flush() {
    return this;
  }

  @Override
  public OutputBuffer<T> fork(Object condition) {
    return this;
  }

  @Override
  public abstract OutputBuffer<T> settings(OutputSettings settings);

  @Override
  public OutputBuffer<T> clone() {
    throw new UnsupportedOperationException();
  }

  private static OutputBuffer<Object> full;

  private static OutputBuffer<Object> done;

  /**
   * Returns an {@code OutputBuffer} in the <em>full</em> state, that binds a
   * {@code null} result.
   */
  @SuppressWarnings("unchecked")
  public static <T> OutputBuffer<T> full() {
    if (full == null) {
      full = new OutputBufferFull<Object>(null, OutputSettings.standard());
    }
    return (OutputBuffer<T>) full;
  }

  /**
   * Returns an {@code OutputBuffer} in the <em>full</em> state, with the given
   * {@code settings}.
   */
  public static <T> OutputBuffer<T> full(OutputSettings settings) {
    if (settings == OutputSettings.standard()) {
      return full();
    }
    return new OutputBufferFull<T>(null, settings);
  }

  /**
   * Returns an {@code OutputBuffer} in the <em>full</em> state, that binds the
   * given {@code value}.
   */
  public static <T> OutputBuffer<T> full(T value) {
    if (value == null) {
      return full();
    }
    return new OutputBufferFull<T>(value, OutputSettings.standard());
  }

  /**
   * Returns an {@code OutputBuffer} in the <em>full</em> state, that binds the
   * given {@code value}, with the given {@code settings}.
   */
  public static <T> OutputBuffer<T> full(T value, OutputSettings settings) {
    if (value == null && settings == OutputSettings.standard()) {
      return full();
    }
    return new OutputBufferFull<T>(value, settings);
  }

  /**
   * Returns an {@code OutputBuffer} in the <em>done</em> state, that binds a
   * {@code null} result.
   */
  @SuppressWarnings("unchecked")
  public static <T> OutputBuffer<T> done() {
    if (done == null) {
      done = new OutputBufferDone<Object>(null, OutputSettings.standard());
    }
    return (OutputBuffer<T>) done;
  }

  /**
   * Returns an {@code OutputBuffer} in the <em>done</em> state, with the given
   * {@code settings}.
   */
  public static <T> OutputBuffer<T> done(OutputSettings settings) {
    if (settings == OutputSettings.standard()) {
      return done();
    }
    return new OutputBufferDone<T>(null, settings);
  }

  /**
   * Returns an {@code OutputBuffer} in the <em>done</em> state, that binds the
   * given {@code value}.
   */
  public static <T> OutputBuffer<T> done(T value) {
    if (value == null) {
      return done();
    }
    return new OutputBufferDone<T>(value, OutputSettings.standard());
  }

  /**
   * Returns an {@code OutputBuffer} in the <em>done</em> state, that binds the
   * given {@code value}, with the given {@code settings}.
   */
  public static <T> OutputBuffer<T> done(T value, OutputSettings settings) {
    if (value == null && settings == OutputSettings.standard()) {
      return done();
    }
    return new OutputBufferDone<T>(value, settings);
  }

  /**
   * Returns an {@code OutputBuffer} in the <em>error</em> state, with the
   * given output {@code error}.
   */
  public static <T> OutputBuffer<T> error(Throwable error) {
    return new OutputBufferError<T>(error, OutputSettings.standard());
  }

  /**
   * Returns an {@code OutputBuffer} in the <em>error</em> state, with the
   * given output {@code error} and {@code settings}.
   */
  public static <T> OutputBuffer<T> error(Throwable error, OutputSettings settings) {
    return new OutputBufferError<T>(error, settings);
  }
}

final class OutputBufferFull<T> extends OutputBuffer<T> {
  final T value;
  final OutputSettings settings;

  OutputBufferFull(T value, OutputSettings settings) {
    this.value = value;
    this.settings = settings;
  }

  @Override
  public boolean isCont() {
    return false;
  }

  @Override
  public boolean isFull() {
    return true;
  }

  @Override
  public boolean isDone() {
    return false;
  }

  @Override
  public boolean isError() {
    return false;
  }

  @Override
  public boolean isPart() {
    return true;
  }

  @Override
  public OutputBuffer<T> isPart(boolean isPart) {
    if (isPart) {
      return OutputBuffer.done(this.value, this.settings);
    } else {
      return this;
    }
  }

  @Override
  public int index() {
    return 0;
  }

  @Override
  public OutputBuffer<T> index(int index) {
    if (index == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid index"), this.settings);
    }
  }

  @Override
  public int limit() {
    return 0;
  }

  @Override
  public OutputBuffer<T> limit(int limit) {
    if (limit == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid limit"), this.settings);
    }
  }

  @Override
  public int capacity() {
    return 0;
  }

  @Override
  public int remaining() {
    return 0;
  }

  @Override
  public byte[] array() {
    throw new UnsupportedOperationException();
  }

  @Override
  public int arrayOffset() {
    throw new UnsupportedOperationException();
  }

  @Override
  public boolean has(int index) {
    return false;
  }

  @Override
  public int get(int index) {
    throw new OutputException();
  }

  @Override
  public void set(int index, int token) {
    throw new OutputException();
  }

  @Override
  public int write(ReadableByteChannel channel) throws IOException {
    return 0;
  }

  @Override
  public OutputBuffer<T> write(int token) {
    return OutputBuffer.error(new OutputException("full"), this.settings);
  }

  @Override
  public OutputBuffer<T> write(String string) {
    return OutputBuffer.error(new OutputException("full"), this.settings);
  }

  @Override
  public OutputBuffer<T> writeln(String string) {
    return OutputBuffer.error(new OutputException("full"), this.settings);
  }

  @Override
  public OutputBuffer<T> writeln() {
    return OutputBuffer.error(new OutputException("full"), this.settings);
  }

  @Override
  public OutputBuffer<T> move(int fromIndex, int toIndex, int length) {
    if (fromIndex == 0 && toIndex == 0 && length == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid move"), this.settings);
    }
  }

  @Override
  public OutputBuffer<T> step(int offset) {
    if (offset == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid step"), this.settings);
    }
  }

  @Override
  public T bind() {
    return this.value;
  }

  @Override
  public OutputSettings settings() {
    return this.settings;
  }

  @Override
  public OutputBuffer<T> settings(OutputSettings settings) {
    return OutputBuffer.full(this.value, settings);
  }

  @Override
  public OutputBuffer<T> clone() {
    return this;
  }
}

final class OutputBufferDone<T> extends OutputBuffer<T> {
  final T value;
  final OutputSettings settings;

  OutputBufferDone(T value, OutputSettings settings) {
    this.value = value;
    this.settings = settings;
  }

  @Override
  public boolean isCont() {
    return false;
  }

  @Override
  public boolean isFull() {
    return false;
  }

  @Override
  public boolean isDone() {
    return true;
  }

  @Override
  public boolean isError() {
    return false;
  }

  @Override
  public boolean isPart() {
    return false;
  }

  @Override
  public OutputBuffer<T> isPart(boolean isPart) {
    if (isPart) {
      return this;
    } else {
      return OutputBuffer.full(this.value, this.settings);
    }
  }

  @Override
  public int index() {
    return 0;
  }

  @Override
  public OutputBuffer<T> index(int index) {
    if (index == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid index"), this.settings);
    }
  }

  @Override
  public int limit() {
    return 0;
  }

  @Override
  public OutputBuffer<T> limit(int limit) {
    if (limit == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid limit"), this.settings);
    }
  }

  @Override
  public int capacity() {
    return 0;
  }

  @Override
  public int remaining() {
    return 0;
  }

  @Override
  public byte[] array() {
    throw new UnsupportedOperationException();
  }

  @Override
  public int arrayOffset() {
    throw new UnsupportedOperationException();
  }

  @Override
  public boolean has(int index) {
    return false;
  }

  @Override
  public int get(int index) {
    throw new OutputException();
  }

  @Override
  public void set(int index, int token) {
    throw new OutputException();
  }

  @Override
  public int write(ReadableByteChannel channel) throws IOException {
    return 0;
  }

  @Override
  public OutputBuffer<T> write(int token) {
    return OutputBuffer.error(new OutputException("done"), this.settings);
  }

  @Override
  public OutputBuffer<T> write(String string) {
    return OutputBuffer.error(new OutputException("done"), this.settings);
  }

  @Override
  public OutputBuffer<T> writeln(String string) {
    return OutputBuffer.error(new OutputException("done"), this.settings);
  }

  @Override
  public OutputBuffer<T> writeln() {
    return OutputBuffer.error(new OutputException("done"), this.settings);
  }

  @Override
  public OutputBuffer<T> move(int fromIndex, int toIndex, int length) {
    if (fromIndex == 0 && toIndex == 0 && length == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid move"), this.settings);
    }
  }

  @Override
  public OutputBuffer<T> step(int offset) {
    if (offset == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid step"), this.settings);
    }
  }

  @Override
  public T bind() {
    return this.value;
  }

  @Override
  public OutputSettings settings() {
    return this.settings;
  }

  @Override
  public OutputBuffer<T> settings(OutputSettings settings) {
    return OutputBuffer.done(this.value, settings);
  }

  @Override
  public OutputBuffer<T> clone() {
    return this;
  }
}

final class OutputBufferError<T> extends OutputBuffer<T> {
  final Throwable error;
  final OutputSettings settings;

  OutputBufferError(Throwable error, OutputSettings settings) {
    this.error = error;
    this.settings = settings;
  }

  @Override
  public boolean isCont() {
    return false;
  }

  @Override
  public boolean isFull() {
    return false;
  }

  @Override
  public boolean isDone() {
    return false;
  }

  @Override
  public boolean isError() {
    return true;
  }

  @Override
  public boolean isPart() {
    return false;
  }

  @Override
  public OutputBuffer<T> isPart(boolean isPart) {
    return this;
  }

  @Override
  public int index() {
    return 0;
  }

  @Override
  public OutputBuffer<T> index(int index) {
    if (index == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid index"), this.settings);
    }
  }

  @Override
  public int limit() {
    return 0;
  }

  @Override
  public OutputBuffer<T> limit(int limit) {
    if (limit == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid limit"), this.settings);
    }
  }

  @Override
  public int capacity() {
    return 0;
  }

  @Override
  public int remaining() {
    return 0;
  }

  @Override
  public byte[] array() {
    throw new UnsupportedOperationException();
  }

  @Override
  public int arrayOffset() {
    throw new UnsupportedOperationException();
  }

  @Override
  public boolean has(int index) {
    return false;
  }

  @Override
  public int get(int index) {
    throw new OutputException();
  }

  @Override
  public void set(int index, int token) {
    throw new OutputException();
  }

  @Override
  public int write(ReadableByteChannel channel) throws IOException {
    return 0;
  }

  @Override
  public OutputBuffer<T> write(int token) {
    return this;
  }

  @Override
  public OutputBuffer<T> write(String string) {
    return this;
  }

  @Override
  public OutputBuffer<T> writeln(String string) {
    return this;
  }

  @Override
  public OutputBuffer<T> writeln() {
    return this;
  }

  @Override
  public OutputBuffer<T> move(int fromIndex, int toIndex, int length) {
    if (fromIndex == 0 && toIndex == 0 && length == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid move"), this.settings);
    }
  }

  @Override
  public OutputBuffer<T> step(int offset) {
    if (offset == 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid step"), this.settings);
    }
  }

  @Override
  public T bind() {
    return null;
  }

  @Override
  public Throwable trap() {
    return this.error;
  }

  @Override
  public OutputSettings settings() {
    return this.settings;
  }

  @Override
  public OutputBuffer<T> settings(OutputSettings settings) {
    return OutputBuffer.error(this.error, settings);
  }

  @Override
  public OutputBuffer<T> clone() {
    return this;
  }
}

File: swim-system-js/swim-core-js/@swim/codec/main/OutputBuffer.ts
// Copyright 2015-2020 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {AnyOutputSettings, OutputSettings} from "./OutputSettings";
import {Format} from "./Format";
import {OutputException} from "./OutputException";
import {Output} from "./Output";

/**
 * Non-blocking token stream buffer.
 */
export abstract class OutputBuffer<T = unknown> extends Output<T> {
  abstract isPart(): boolean;
  abstract isPart(isPart: boolean): OutputBuffer<T>;

  abstract index(): number;
  abstract index(index: number): OutputBuffer<T>;

  abstract limit(): number;
  abstract limit(limit: number): OutputBuffer<T>;

  abstract capacity(): number;

  abstract remaining(): number;

  abstract has(index: number): boolean;

  abstract get(index: number): number;

  abstract set(index: number, token: number): void;

  abstract write(token: number): OutputBuffer<T>;
  abstract write(string: string): OutputBuffer<T>;

  writeln(string?: string): OutputBuffer<T> {
    if (typeof string === "string") {
      this.write(string);
    }
    return this.write(this.settings().lineSeparator());
  }

  display(object: unknown): OutputBuffer<T> {
    Format.display(object, this);
    return this;
  }

  debug(object: unknown): OutputBuffer<T> {
    Format.debug(object, this);
    return this;
  }

  abstract step(offset?: number): OutputBuffer<T>;

  flush(): OutputBuffer<T> {
    return this;
  }

  abstract settings(): OutputSettings;
  abstract settings(settings: AnyOutputSettings): Output<T>;

  clone(): OutputBuffer<T> {
    throw new Error();
  }

  private static _fullBuffer?: OutputBuffer<any>;

  private static _doneBuffer?: OutputBuffer<any>;

  /**
   * Returns an `OutputBuffer` in the _full_ state, that binds a `null` result.
   */
  static full<T>(value: T | null = null, settings: OutputSettings = OutputSettings.standard()): OutputBuffer<T> {
    if (value === null && settings === OutputSettings.standard()) {
      if (OutputBuffer._fullBuffer === void 0) {
        OutputBuffer._fullBuffer = new OutputBufferFull(value, OutputSettings.standard());
      }
      return OutputBuffer._fullBuffer;
    }
    return new OutputBufferFull<T>(value!, settings);

  }

  /**
   * Returns an `OutputBuffer` in the _done_ state, that binds a `null` result.
   */
  static done<T>(value: T | null = null, settings: OutputSettings = OutputSettings.standard()): OutputBuffer<T> {
    if (value === null && settings === OutputSettings.standard()) {
      if (OutputBuffer._doneBuffer === void 0) {
        OutputBuffer._doneBuffer = new OutputBufferDone(value, OutputSettings.standard());
      }
      return OutputBuffer._doneBuffer;
    }
    return new OutputBufferDone<T>(value!, settings);
  }

  /**
   * Returns an `OutputBuffer` in the _error_ state, with the given output `error`.
   */
  static error<T>(error: Error, settings: OutputSettings = OutputSettings.standard()): OutputBuffer<T> {
    return new OutputBufferError<T>(error, settings);
  }
}

/** @hidden */
class OutputBufferFull<T> extends OutputBuffer<T> {
  /** @hidden */
  readonly _value: T;
  /** @hidden */
  readonly _settings: OutputSettings;

  constructor(value: T, settings: OutputSettings) {
    super();
    this._value = value;
    this._settings = settings;
  }

  isCont(): boolean {
    return false;
  }

  isFull(): boolean {
    return true;
  }

  isDone(): boolean {
    return false;
  }

  isError(): boolean {
    return false;
  }

  isPart(): boolean;
  isPart(isPart: boolean): OutputBuffer<T>;
  isPart(isPart?: boolean): boolean | OutputBuffer<T> {
    if (isPart === void 0) {
      return true;
    } else if (isPart) {
      return OutputBuffer.done(this._value, this._settings);
    } else {
      return this;
    }
  }

  index(): number;
  index(index: number): OutputBuffer<T>;
  index(index?: number): number | OutputBuffer<T> {
    if (index === void 0) {
      return 0;
    } else if (index === 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid index"), this._settings);
    }
  }

  limit(): number;
  limit(limit: number): OutputBuffer<T>;
  limit(limit?: number): number | OutputBuffer<T> {
    if (limit === void 0) {
      return 0;
    } else if (limit === 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid index"), this._settings);
    }
  }

  capacity(): number {
    return 0;
  }

  remaining(): number {
    return 0;
  }

  has(index: number): boolean {
    return false;
  }

  get(index: number): number {
    throw new OutputException();
  }

  set(index: number, token: number): void {
    throw new OutputException();
  }

  write(token: number): OutputBuffer<T>;
  write(string: string): OutputBuffer<T>;
  write(tokenOrString: number | string): OutputBuffer<T> {
    return OutputBuffer.error(new OutputException("full"), this._settings);
  }

  writeln(string?: string): OutputBuffer<T> {
    return OutputBuffer.error(new OutputException("full"), this._settings);
  }

  step(offset: number = 1): OutputBuffer<T> {
    if (offset === 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid step"), this._settings);
    }
  }

  bind(): T {
    return this._value;
  }

  settings(): OutputSettings;
  settings(settings: OutputSettings): OutputBuffer<T>;
  settings(settings?: OutputSettings): OutputSettings | OutputBuffer<T> {
    if (settings === void 0) {
      return this._settings;
    } else {
      return OutputBuffer.full(this._value, settings);
    }
  }

  clone(): OutputBuffer<T> {
    return this;
  }
}

/** @hidden */
class OutputBufferDone<T> extends OutputBuffer<T> {
  /** @hidden */
  readonly _value: T;
  /** @hidden */
  readonly _settings: OutputSettings;

  constructor(value: T, settings: OutputSettings) {
    super();
    this._value = value;
    this._settings = settings;
  }

  isCont(): boolean {
    return false;
  }

  isFull(): boolean {
    return false;
  }

  isDone(): boolean {
    return true;
  }

  isError(): boolean {
    return false;
  }

  isPart(): boolean;
  isPart(isPart: boolean): OutputBuffer<T>;
  isPart(isPart?: boolean): boolean | OutputBuffer<T> {
    if (isPart === void 0) {
      return false;
    } else if (isPart) {
      return this;
    } else {
      return OutputBuffer.full(this._value, this._settings);
    }
  }

  index(): number;
  index(index: number): OutputBuffer<T>;
  index(index?: number): number | OutputBuffer<T> {
    if (index === void 0) {
      return 0;
    } else if (index === 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid index"), this._settings);
    }
  }

  limit(): number;
  limit(limit: number): OutputBuffer<T>;
  limit(limit?: number): number | OutputBuffer<T> {
    if (limit === void 0) {
      return 0;
    } else if (limit === 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid index"), this._settings);
    }
  }

  capacity(): number {
    return 0;
  }

  remaining(): number {
    return 0;
  }

  has(index: number): boolean {
    return false;
  }

  get(index: number): number {
    throw new OutputException();
  }

  set(index: number, token: number): void {
    throw new OutputException();
  }

  write(token: number): OutputBuffer<T>;
  write(string: string): OutputBuffer<T>;
  write(tokenOrString: number | string): OutputBuffer<T> {
    return OutputBuffer.error(new OutputException("done"), this._settings);
  }

  writeln(string?: string): OutputBuffer<T> {
    return OutputBuffer.error(new OutputException("done"), this._settings);
  }

  step(offset: number = 1): OutputBuffer<T> {
    if (offset === 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid step"), this._settings);
    }
  }

  bind(): T {
    return this._value;
  }

  settings(): OutputSettings;
  settings(settings: OutputSettings): OutputBuffer<T>;
  settings(settings?: OutputSettings): OutputSettings | OutputBuffer<T> {
    if (settings === void 0) {
      return this._settings;
    } else {
      return OutputBuffer.done(this._value, settings);
    }
  }

  clone(): OutputBuffer<T> {
    return this;
  }
}

/** @hidden */
class OutputBufferError<T> extends OutputBuffer<T> {
  /** @hidden */
  readonly _error: Error;
  /** @hidden */
  readonly _settings: OutputSettings;

  constructor(error: Error, settings: OutputSettings) {
    super();
    this._error = error;
    this._settings = settings;
  }

  isCont(): boolean {
    return false;
  }

  isFull(): boolean {
    return false;
  }

  isDone(): boolean {
    return false;
  }

  isError(): boolean {
    return true;
  }

  isPart(): boolean;
  isPart(isPart: boolean): OutputBuffer<T>;
  isPart(isPart?: boolean): boolean | OutputBuffer<T> {
    if (isPart === void 0) {
      return false;
    } else {
      return this;
    }
  }

  index(): number;
  index(index: number): OutputBuffer<T>;
  index(index?: number): number | OutputBuffer<T> {
    if (index === void 0) {
      return 0;
    } else if (index === 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid index"), this._settings);
    }
  }

  limit(): number;
  limit(limit: number): OutputBuffer<T>;
  limit(limit?: number): number | OutputBuffer<T> {
    if (limit === void 0) {
      return 0;
    } else if (limit === 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid index"), this._settings);
    }
  }

  capacity(): number {
    return 0;
  }

  remaining(): number {
    return 0;
  }

  has(index: number): boolean {
    return false;
  }

  get(index: number): number {
    throw new OutputException();
  }

  set(index: number, token: number): void {
    throw new OutputException();
  }

  write(token: number): OutputBuffer<T>;
  write(string: string): OutputBuffer<T>;
  write(tokenOrString: number | string): OutputBuffer<T> {
    return this;
  }

  writeln(string?: string): OutputBuffer<T> {
    return this;
  }

  step(offset: number = 1): OutputBuffer<T> {
    if (offset === 0) {
      return this;
    } else {
      return OutputBuffer.error(new OutputException("invalid step"), this._settings);
    }
  }

  bind(): T {
    throw new OutputException();
  }

  trap(): Error {
    return this._error;
  }

  settings(): OutputSettings;
  settings(settings: OutputSettings): OutputBuffer<T>;
  settings(settings?: OutputSettings): OutputSettings | OutputBuffer<T> {
    if (settings === void 0) {
      return this._settings;
    } else {
      return OutputBuffer.error(this._error, settings);
    }
  }

  clone(): OutputBuffer<T> {
    return this;
  }
}
